Feat: Implement Offline Caching for Movie Lists & Add Bookmark Flow

yusufnasserdev opened the pull request:
# Feat: Implement Offline Caching for Movie Lists & Add Bookmark Flow

This branch started with the simple goal of implementing a simple sheet, but this was a journey, and not a simple one.

This pull request introduces two major features that significantly enhance the user experience: a robust offline caching mechanism for custom movie lists and a new "Save to List" bottom sheet for bookmarking movies.

---

### 1. Offline Caching and Background Sync for Movie Lists

To provide a seamless, offline-first experience, the application now caches all user-created movie lists and their contents. The data layer has been re-architected to prioritize displaying cached data, ensuring instant loads and availability without a network connection.

**Key Implementation Details:**
* **Background Sync with `WorkManager`**: A new `MovieListSyncWorker` has been introduced to periodically refresh the local cache in the background. This ensures that the local data stays reasonably up-to-date without blocking the user. The application class now provides the necessary `HiltWorkerFactory` and schedules the periodic work.
* **Extended Room Database**: The local `NovixDatabase` has been expanded with three new tables:
    * `cached_movie_lists`
    * `movie_list_membership`
    * `sync_metadata`
* **Cache-First Repository**: The `CustomMovieListRepositoryImpl` is now fundamentally cache-first.
    * It fetches data from the `CustomMovieListLocalDataSource` first.
    * If the cache is stale or a force refresh is requested, it fetches fresh data from the network, updates the local cache, and then returns the data.
    * A `Mutex` is used to prevent race conditions during sync operations.
* **New Data Layer Components**: New DAOs (`MovieListDao`, `ListMembershipDao`, `SyncMetadataDao`) and a `CustomMovieListLocalDataSource` have been created to abstract database operations. These are all integrated via Hilt dependency injection.

---

### 2. "Save to List" Bookmark Flow

A new, reusable bottom sheet component, `BookmarkBottomSheet`, has been created to allow users to easily save movies to their custom lists.

**Key Features & Implementation:**
* **User-Friendly UI**: The bottom sheet displays a list of available movie lists where a user can add the selected movie. It also provides a clear entry point to create a new list.
* **State Management**: The sheet is driven by `BookmarkSheetViewModel`, which handles:
    * Fetching the lists a movie can be added to (`GetAvailableListsForMovie` use case).
    * Managing loading, empty, and success states.
    * Handling the logic for adding a movie to multiple selected lists.
* **Guest User Handling**: If the user is not logged in, the sheet displays a prompt to log in before they can use the functionality. The ViewModel checks the user's login status to control this state.
* **Seamless Integration**:
    * The feature is integrated into the Home screen's "Popular," "Top-Rated," "Upcoming," and "Continue Watching" sections, replacing placeholder save icons.
    * A `LocalNavController` is provided via `CompositionLocalProvider` to enable navigation from the bottom sheet (e.g., to the Login screen or the Create List screen).

---

### 3. Major Refactoring and Other Improvements

* **`UInt` to `Int` Refactoring**: The experimental `UInt` type has been replaced with `Int` for all identifiers (`listId`, `movieId`) across the domain, data, and presentation layers. This improves stability and interoperability with Java.
* **`MediaType` Enum Enhancement**: The `MediaType` enum no longer uses an integer representation. It now includes `isMovie()` and `isTvShow()` extension functions for improved type safety and readability. This change has been propagated through the movie/TV show details and reviews screens.
* **Navigation with Arguments**: `Screen.Lists` is now a `@Serializable` data class that accepts a `createList` boolean argument. This allows direct navigation to the lists screen with the "create new list" bottom sheet already open.
* **UI & String Updates**:
    * Various string resources have been added and corrected for clarity and to support the new features.
    * Corrected a typo in a content description from `imagr_dot` to `image_dot`.
    * The layout of the `Selection` component was changed from a `Row` to a `Column`.

## Development

Closes #585 In LondonSquad/Novix

# Note to Reviewers

Tests are still being worked on, considering the size of the PR, I thought to create it so you can review the production code and test the app while I work on the tests.

yusufnasserdev committed:
Refactor: Update cancel icon color in Design System

The fill color of the cancel icon drawable has been changed from `#1F1F1F` with an alpha of `0.87` to `#000000` with an alpha of `1.00`.

yusufnasserdev committed:
feat: add string resources & correct typos

yusufnasserdev committed:
feat: add `isNotVisible` property to `BottomSheetState`

yusufnasserdev committed:
feat: Add `BookmarkSheetUiState`

yusufnasserdev committed:
feat: define `BookmarkSheetContract` interface

yusufnasserdev committed:
feat: define BookmarkSheetEffect sealed interface

yusufnasserdev committed:
feat: Implement basic bookmark sheet ViewModel

Further implementation for fetching lists and integrating with domain/data layers is pending.

yusufnasserdev committed:
feat: add initial bookmark bottom sheet implementation

yusufnasserdev committed:
Merge branch 'develop' into feature/add-bookmark-sheet

# Conflicts:
#	presentation/src/main/res/values-ar/strings.xml
#	presentation/src/main/res/values/strings.xml

yusufnasserdev committed:
Refactor: move bookmark sheet to shared directory and implement list functionality

yusufnasserdev committed:
Merge branch 'develop' into feature/add-bookmark-sheet

# Conflicts:
#	presentation/src/main/res/values/strings.xml

yusufnasserdev committed:
feat: add strings for save to list, create new list, and success/error messages

yusufnasserdev committed:
fix: update selection component and bookmark sheet interaction

yusufnasserdev committed:
refactor: remove specific item addition effects from BookmarkSheetEffect

yusufnasserdev committed:
feat: implement GetMovieListsUseCase and repository method

yusufnasserdev committed:
feat: add success/error snackbars to bookmark sheet

yusufnasserdev committed:
refactor: update string resources for item count and list creation

yusufnasserdev committed:
refactor: use PagingData for lists in BookmarkSheet

yusufnasserdev committed:
refactor: standardize snackbar messages and add mapper for bookmark sheet

yusufnasserdev committed:
feat: add empty getMovieLists to MovieDetailsRepositoryImpl

yusufnasserdev committed:
feat: implement manage bookmark functionality in home screen

yusufnasserdev committed:
Merge branch 'develop' into feature/add-bookmark-sheet

# Conflicts:
#	data/src/main/java/com/london/data/remote/service/details/movie/MovieDetailsApiService.kt
#	domain/src/main/java/com/london/domain/repository/MovieDetailsRepository.kt
#	presentation/src/main/java/com/london/presentation/feature/home/HomeScreenContract.kt
#	presentation/src/main/java/com/london/presentation/feature/home/HomeScreenUiState.kt
#	presentation/src/main/java/com/london/presentation/feature/home/popular/PopularSection.kt
#	presentation/src/main/java/com/london/presentation/feature/list/savedlist/listScreen.kt
#	presentation/src/main/res/values-ar/strings.xml
#	presentation/src/main/res/values/strings.xml

yusufnasserdev committed:
refactor: remove unused onSaveClick from PopularSection

yusufnasserdev committed:
feat: implement item selection and snackbar logic for bookmark sheet

yusufnasserdev committed:
feat: implement save functionality in home screen sections

yusufnasserdev committed:
feat: provide NavController using CompositionLocal

yusufnasserdev committed:
refactor: rename `createNewList` to `createList` in Lists screen

yusufnasserdev committed:
feat: navigate to create list screen from bookmark sheet

yusufnasserdev committed:
feat: add login_to_add_list string

yusufnasserdev committed:
feat: add guest login button drawables

yusufnasserdev committed:
feat: add guest login support to bookmark sheet

yusufnasserdev committed:
feat: open create list sheet when clicking on `Create new list` button.

yusufnasserdev committed:
refactor: improve list and bookmark sheet UI

yusufnasserdev committed:
feat: add "no lists found" message to bookmark sheet

yusufnasserdev committed:
feat: create mapping for movie to lists

yusufnasserdev committed:
feat: preselect lists when bookmark sheet is shown

yusufnasserdev committed:
feat: update bookmark sheet when shown

yusufnasserdev committed:
refactor: update "no lists found" string to "no lists available"

yusufnasserdev committed:
refactor: change toBookmarkUiLists to map List instead of Flow<PagingData>

yusufnasserdev committed:
fix: filter out lists containing the current movie from bookmark sheet

yusufnasserdev committed:
feat: add strings for item added to list status

yusufnasserdev committed:
feat: implement manage bookmark functionality in continue watching section

yusufnasserdev committed:
feat: add success and error snack bars to bookmark sheet

yusufnasserdev committed:
refactor: revert GetAllListedMovies

yusufnasserdev committed:
feat: implement getMovieLists functionality

yusufnasserdev committed:
refactor: update BookmarkSheetViewModel dependencies and logic

yusufnasserdev committed:
refactor: remove getMovieLists functionality

yusufnasserdev committed:
refactor: improve GetAllListedMovies performance by fetching all pages concurrently

yusufnasserdev committed:
fix: reset BookmarkSheet state on dismiss and hide

yusufnasserdev committed:
feat: Implement local caching for custom movie lists

This commit introduces local caching for custom movie lists and their memberships.

Key changes include:
- Added Room entities: `MovieListLocal`, `MovieListMembershipLocal`, and `SyncMetadataLocal`.
- Implemented DAOs: `MovieListDao`, `ListMembershipDao`, and `SyncMetadataDao`.
- Created `CustomMovieListLocalDataSource` and its implementation to manage local data.
- Updated `CustomMovieListRepositoryImpl` to integrate with the local data source, including cache refresh logic and mutex for synchronization.
- Introduced `MovieListSyncWorker` to periodically refresh the cache in the background using WorkManager.
- Configured `NovixApplication` to initialize WorkManager and schedule the periodic sync.
- Updated Hilt modules to provide new DAOs, local data source, and WorkManager.
- Modified `id` and `moviesCount` in `MovieList` entity from `UInt` to `Int`.
- Added necessary dependencies for WorkManager and Hilt.

yusufnasserdev committed:
chore: update Arabic translations for list-related strings

yusufnasserdev committed:
feat: enhance MovieListDao with sorting, pagination, and update operations

yusufnasserdev committed:
feat: enhance ListMembershipDao with pagination and count queries

yusufnasserdev committed:
refactor: improve custom movie list caching and error handling

yusufnasserdev committed:
refactor: update movie list use cases and repository

yusufnasserdev committed:
feat: migrate to signed types in BookmarkUiList and BookmarkBottomSheet

yusufnasserdev committed:
refactor: update ID types from UInt to Int

yusufnasserdev committed:
refactor: improve BookmarkSheet UI and state handling

yusufnasserdev committed:
Merge branch 'develop' into feature/add-bookmark-sheet

# Conflicts:
#	data/src/main/java/com/london/data/di/RepositoryModule.kt
#	data/src/main/java/com/london/data/remote/source/details/movie/MovieDetailsRemoteDataSourceImpl.kt
#	data/src/main/java/com/london/data/repository/list/CustomMovieListRepositoryImpl.kt
#	data/src/main/java/com/london/data/repository/search/MovieDetailsRepositoryImpl.kt
#	data/src/test/java/com/london/data/repository/list/CustomMovieListRepositoryImplTest.kt
#	domain/src/main/java/com/london/domain/entity/recent/MediaType.kt
#	domain/src/main/java/com/london/domain/repository/MovieDetailsRepository.kt
#	presentation/src/main/java/com/london/presentation/feature/details/actor/ActorDetailsScreen.kt
#	presentation/src/main/res/values-ar/strings.xml
#	presentation/src/main/res/values/strings.xml

BassantM commented on app/src/main/java/com/london/app/NovixApplication.kt:
```diff
@@ -1,15 +1,32 @@
 
 @HiltAndroidApp
-class NovixApplication : Application() {
+class NovixApplication() : Application(), Configuration.Provider {
```
remove parameters () to be class NovixApplication instead of class NovixApplication()

yusufnasserdev replied:
Done

BassantM commented on domain/src/main/java/com/london/domain/usecase/movielist/AddMovieToListUseCase.kt:
```diff
@@ -4,9 +4,9 @@ import com.london.domain.repository.CustomMovieListRepository
 import javax.inject.Inject
 
 class AddMovieToListUseCase @Inject constructor(
```
remove this function to manage package

yusufnasserdev replied:
Will be in another PR.

BassantM commented on domain/src/main/java/com/london/domain/usecase/movielist/RemoveMovieFromListUseCase.kt:
```diff
@@ -7,6 +7,6 @@ class RemoveMovieFromListUseCase @Inject constructor(
 ) {
 
-    suspend fun invoke(listId: UInt, movieId: UInt): Boolean =
+    suspend fun invoke(listId: Int, movieId: Int): Boolean =
```
remove into manage movie lists

yusufnasserdev replied:
Will be in another PR.

BassantM commented on domain/src/main/java/com/london/domain/usecase/movielist/GetAvailableListsForMovie.kt:
```diff
@@ -0,0 +1,33 @@
+import com.london.domain.repository.CustomMovieListRepository
+import javax.inject.Inject
+
+class GetAvailableListsForMovie @Inject constructor(
```
i think u can merge all related use cases in one class

yusufnasserdev replied:
Will be in another PR.

BassantM commented on domain/src/main/java/com/london/domain/usecase/movielist/GetMovieListDetailsUseCase.kt:
```diff
@@ -9,6 +9,6 @@ class GetMovieListDetailsUseCase @Inject constructor(
 ) {
 
-    suspend fun invoke(listId: UInt, pageNumber: Int): PagedFetchResponse<Movie> =
+    suspend fun invoke(listId: Int, pageNumber: Int): PagedFetchResponse<Movie> =
```
same

yusufnasserdev replied:
Will be in another PR.

BassantM commented on data/src/main/java/com/london/data/local/source/customLists/CustomMovieListLocalDataSource.kt:
```diff
@@ -0,0 +1,43 @@
+
+    // Movie list metadata queries
+    suspend fun getAllUserLists(): List<MovieListLocal>
+    fun getAllUserListsFlow(): Flow<List<MovieListLocal>>
```
I wanna ask why some functions are suspended and othersare  not

yusufnasserdev replied:
`suspend` is designed for functions that perform a single asynchronous operation and return once.

Flows are reactive streams, that emit values over time, when you observe a Flow, you're already working asynchronously, no need to suspend.

BassantM commented on presentation/src/main/java/com/london/presentation/feature/home/trending/tvshow/TrendingTvShowsScreen.kt:
```diff
@@ -51,7 +51,7 @@ fun TrendingTvShowsScreen(
         onBack = viewModel::onBack,
         onRetry = viewModel::onRetry,
-        emptyLayoutMessage = R.string.no_trending_tvshows_in_genre,
+        emptyLayoutMessage = R.string.no_trending_shows_in_genre,
```
Why did u rename this? 
This text is from the Figma design

yusufnasserdev replied:
It's just a simple rename, `shows` are clearer than `tvshows`, it was showing a linting error.

BassantM commented on presentation/src/main/res/values/strings.xml:
```diff
@@ -100,6 +100,7 @@
     <string name="no_trending_movies_in_genre">No movies found in this genre.</string>
     <string name="no_rating_items_in_list">No items found in my rating list.</string>
     <string name="no_trending_tvshows_in_genre">No tv shows found in this genre.</string>
```
by replacing line 103 by 102 why didn't delete it

yusufnasserdev replied:
I will check this.

BassantM commented on presentation/src/main/java/com/london/presentation/shared/bookmarkSheet/BookmarkSheetViewModel.kt:
```diff
@@ -0,0 +1,129 @@
+        tryToExecute(
+            onStart = { updateState { copy(isAddingToList = true) } },
+            block = {
+                listsToAdd.forEach { listId ->
+                    addMovieToListUseCase.invoke(
+                        listId = listId,
+                        movieId = bookmarkedId
```
extract this

BassantM commented on data/src/main/java/com/london/data/worker/MovieListSyncWorker.kt:
```diff
@@ -0,0 +1,97 @@
+            )
+        }
+
+        fun syncNow(workManager: WorkManager) {
```
was not used

yusufnasserdev committed:
refactor: clean up code and remove unused string

yusufnasserdev committed:
feat: Add CrashReporter to CustomMovieListRepositoryImpl and update tests

yusufnasserdev committed:
refactor: improve CustomMovieListRepositoryImpl tests

yusufnasserdev committed:
test: update GetAllListedMoviesTest

yusufnasserdev committed:
Merge branch 'develop' into feature/add-bookmark-sheet

# Conflicts:
#	data/src/test/java/com/london/data/repository/list/CustomMovieListRepositoryImplTest.kt
#	domain/src/main/java/com/london/domain/usecase/movielist/AddMovieToListUseCase.kt
#	domain/src/main/java/com/london/domain/usecase/movielist/GetAllListedMovies.kt
#	domain/src/main/java/com/london/domain/usecase/movielist/GetMovieListDetailsUseCase.kt
#	domain/src/main/java/com/london/domain/usecase/movielist/GetMovieListNameUseCase.kt
#	domain/src/main/java/com/london/domain/usecase/movielist/ManageMovieListUseCase.kt
#	domain/src/main/java/com/london/domain/usecase/movielist/RemoveMovieFromListUseCase.kt
#	domain/src/test/kotlin/com/london/domain/usecase/movielist/AddMovieToListUseCaseTest.kt
#	domain/src/test/kotlin/com/london/domain/usecase/movielist/GetAllMovieListsUseCaseTest.kt
#	domain/src/test/kotlin/com/london/domain/usecase/movielist/GetMovieListDetailsUseCaseTest.kt
#	domain/src/test/kotlin/com/london/domain/usecase/movielist/GetMovieListNameUseCaseTest.kt
#	domain/src/test/kotlin/com/london/domain/usecase/movielist/ManageMovieListUseCaseTest.kt
#	domain/src/test/kotlin/com/london/domain/usecase/movielist/RemoveMovieFromListUseCaseTest.kt
#	presentation/src/main/java/com/london/presentation/feature/home/HomeScreen.kt
#	presentation/src/main/java/com/london/presentation/feature/home/HomeScreenContract.kt
#	presentation/src/main/java/com/london/presentation/feature/home/HomeScreenUiState.kt
#	presentation/src/main/java/com/london/presentation/feature/home/HomeViewModel.kt
#	presentation/src/main/java/com/london/presentation/feature/home/section/CarouselSection.kt
#	presentation/src/main/java/com/london/presentation/feature/list/viewitems/ViewItemsViewModel.kt
#	presentation/src/main/java/com/london/presentation/feature/reviews/ReviewsViewModel.kt

london-bot[bot] committed:
Branch was auto-updated.

yusufnasserdev committed:
refactor: comment out TopTvShowsPicksViewModelTest `onRetryClick should clear error state and fetch data again` test

yusufnasserdev committed:
refactor: Initialize WorkManager on demand

This commit updates the WorkManager initialization to be on-demand.
This involves:
- Disabling the default `WorkManagerInitializer` in `AndroidManifest.xml`.
- Manually initializing WorkManager in `NovixApplication.kt` using `WorkManager.initialize()`.

yusufnasserdev committed:
chore: keep the names of Hilt-injected Workers

yusufnasserdev committed:
refactor: comment out guest `onLoginAsGuestClick with success emits NavigateToHome`
